doc/flowvisor.8
.ds PN flowvisor

.TH flowvisor 8 "May 2008" "FlowVisor" "FlowVisor Manual"

.SH NAME
flowvisor \- OpenFlow network slicing daemon

.SH AUTHOR

Rob Sherwood <rob.sherwood@stanford.edu>

Ali Al-Shabibi <ali.al-shabibi@stanford.edu>

.SH SYNOPSIS
.B flowvisor
[-d \fIdebug\fR] [-h] [-l] [-p \fIport\fR] [\fIconfig.json\fR] 

.SH DESCRIPTION
The
.B flowvisor
program is a daemon that logically slices an OpenFlow-enabled switch into multiple, logical switches each controlled by a different OpenFlow controller. 

For more details, see:

"Can the Production Network Be the Test-bed?" 
Rob Sherwood, Glen Gibb, Kok-Kiong Yap, Guido Appenzeller, Martin Casado, Nick McKeown, Guru Parulkar.
Symposium for Operating Systems Design and Implementation, 2010

.SH OPTIONS

.TP
-d \fIdebug\fR
Overrides debugging threshold in config.json and sets it to \fIdebug\fR.  Does not modify config.json.
Mostly used for manual debugging.  The string \fIdebug\fR must be one of MOBUG, DEBUG, NOTE, INFO, WARN, ALERT, CRIT,
and FATAL (in order of increasing priority/verbosity) similar to syslog(3).

.TP
-h
Prints the version number and usage information and exits.


.TP
-l
Logs to stderr instead of to syslog()

.TP
-p \fIport\fR
Overrides listenning port from config.json and sets it to \fIport\fR.  Does not modify config.json.
Mostly used for manual debugging.

.TP
\fIconfig.json\fR
Supplying the input config file will cause Flowvisor to overwrite the config in its embedded database with whatever is in the config file. 

.SH ENVIRONMENT

.TP 
FV_JVM_ARGS

The override the default arguments to the JVM with those in $FV_JVM_ARGS.
May be useful for increasing memory efficiency or playing with special
JVM optimizations, like fancy garbage collectors.

.TP
FV_DEBUG_PORT

Turns on remote debugging support for the JVM and sets the JVM to
listen on port $FV_DEBUG_PORT.  This is useful for attaching a
remote debugger, e.g., `jdb -attach fvserver.foo.com:$FV_DEBUG_PORT`.
\fINOTE\fR: there is no security in the JVM debugging protocol, so
by enabling this option you potentially open your running FlowVisor
to remote control/manipulation.  FV_DEBUG_PORT is specified as
\fI[local_address:]port\fR, where \fIlocal_address\fR defaults to "all
interfaces", so if you are interested in binding to a specific
interface, e.g., the loopback interface on port 12345, can set
FV_DEBUG_PORT=localhost:12345 .  See
\fIhttp://download.oracle.com/javase/6/docs/technotes/guides/jpda/conninv.html#Invocation\fR
for more information.



.SH "SEE ALSO"

.BR fvctl (1)
.BR fvconfig (1)


RELEASE-NOTES
Flowvisor 0.8.6 : September 30 2012
    * FLOWVISOR-43   FlowVisor 0.6.x package improvements
    * FLOWVISOR-90   Nightly build packages in unstable apt repo
    * FLOWVISOR-125  /etc/init.d/flowvisor variable instantiation (?) problems in 0.8.1
    * FLOWVISOR-136  Path and naming mismatches between FlowVisor package and documentation
    * FLOWVISOR-137  Missing instructions for package install
    * FLOWVISOR-140  Add /etc/init.d/flowvisor status
    * FLOWVISOR-149  Create package installation user if it doesn't exist
    * FLOWVISOR-30   make "port alread bound" fatal error clearer to the user
    * FLOWVISOR-34   new passwd prompt gets NPE if non-terminal
    * FLOWVISOR-181  "fvctl changePasswd" doesn't change password

Flowvisor 0.8.5 : July 11 2012
    * FLOWVISOR-177: fvctl dumpConfig should output the config file in fvctl's working directory
    * FLOWVISOR-178: Slice creation renders flowvisor unresponsive in some cases
    * FLOWVISOR-179: Need callback when slice controller disconnects

Flowvisor 0.8.4 : June 22 2012
    * FLOWVISOR-175: Provide API to change slice password.
    * FLOWVISOR-173: FV rejects a Flow_mod request with DELETE_FLOW if the buffer_id is 0.
    * FLOWVISOR-167: Need an RPC API to allow db backup
    * FLOWVISOR-166: Update API to replace setConfig/getConfig
    * FLOWVISOR-165: FlowRule ids change when the flowspace is manipulated
    * FLOWVISOR-161: Checkpointing broken on FlowSpace modifications    
    * FLOWVISOR-160: Prevent NoMatch exception from bubbling up to the output
    * FLOWVISOR-159: Flowvisor crashes on empty flowspaces if there are controller connected

Flowvisor 0.8.3 : May 20 2012
    * Database-based flowspace storage
    * Config storage transitioned to JSON
    * Dump config file via API call
    * FLOWVISOR-104: writing config after each change to disk hangs flowvisors with large flowspace
    
Flowvisor 0.8.2 : Febuary 2 2012
    * Slice configuration parameter for controlling LLDP spam
    * JSON callbacks for fine-grained topology changes
    * FLOWVISOR-128: Warn users when adding flowspace rules for non-existing slices
    * FLOWVISOR-126: Slices containing '<' can not be deleted
    * FLOWVISOR-122: fvctl no longer supports registration of topology callbacks with authorization information
    * FLOWVISOR-121: XMLRPC web server logs to stderr instead of FVLog
    * FLOWVISOR-95: when slicing actions, the errors that bubble up are confusing
    * FLOWVISOR-73: Buffer ID sanitizing
    * FLOWVISOR-22: add config option for drop policy

Flowvisor 0.8.1 : July 22 2011
    * Better logging to find delays and process starvation
    * Changed default GC to be mark/sweep
    * Source documentation improvements
    * FLOWVISOR-120: disabled DNS lookup for stats_desc
    * FLOWVISOR-118: fixed FvSlicer memleak in XID translation
    * FLOWVISOR-117: bind fails for override IPv4 stacks
    * FLOWVISOR-116: move configfile read before logging

FlowVisor 0.8.0 : July 13 2011
    * fixed problem with FlowIDs exhausting
    * fixed memleaks that caused instability
    * fixed dynamic ports and topology discovery issues
    * fixed bugs in Vendor Stats and sendlen handling
    * updated developer documentation
    * added better support for remote debugging and JVM tweaks
    * fixed IPv6 dual-stack local port binding bug
    * changed superuser from 'root' to 'fvadmin'
    * lots of bugs fixed and sanity checks implemented

FlowVisor 0.7.2 : Mar 17 2011
    * added changeSlice fvctl/API command
    * fixed big bug in drop rules for disconnected controllers
    * fixed Null Pointer Exception race condition; stops crashing
    * added MRTG formatting scripts (thanks Chris Tengi!)

FlowVisor 0.7.1 : Feb 22 2011
    * diabled flow tracking by default
    * fixed bug in getSwitchFlowDB

FlowVisor 0.7 : 
    * fixed big I/O bug that caused switch flapping
    * added msg statistics, see `fvctl get{Slice,Switch}stats`
    * added flow tracking, see `fvctl get{Switch,SliceRewrite}FlowDB`
        to enable, run `fvctl setConfig flowvisor\!track_flows true`
    * lots of performance improvements
        * flow setup time ~5x improved
    * lots of small bugs fixed
    * password prompt no longer echos plaintext
    * better error handling in create slice and change flowspace

FlowVisor 0.6 - total rewrite (in Java!)
    * Single, unified flowspace config, not divided by slice
    * lldp is now rewritten with a trailer, not by rewritting the dst mac
    * OFError msgs now correctly encap the offending packet
    * NEW FlowVisor UserAPI: query and manipulate slice info


src/org/flowvisor/config/Flowvisor.java
package org.flowvisor.config;

/**
 * A proxy interface for flowvisor specific 
 * information
 * 
 * 
 * @author ash
 *
 */
public interface Flowvisor extends FVAppConfig {
	
	// COLUMN NAMES
	public static String TRACK = "track_flows";
	public static String STATS = "stats_desc_hack";
	public static String LISTEN = "listen_port";
	public static String APIPORT = "api_webserver_port";
	public static String CHECKPOINT = "checkpointing";
	public static String JETTYPORT = "api_jetty_webserver_port";
	public static String FLOODPERM = "default_flood_perm";
	public static String LOGIDENT = "log_ident";
	public static String LOGGING = "logging";
	public static String LOGFACILITY = "log_facility";
	public static String TOPO = "run_topology_server";
	public static String VERSION = "version";
	public static String HOST = "host";
	public static String CONFIG = "config_name";
	
	// Table name
	public static String FLOWVISOR = "flowvisor";
	
	public Boolean gettrack_flows(Integer id) throws ConfigError;
	public Boolean gettrack_flows() throws ConfigError;
	public Boolean getstats_desc_hack(Integer id);
	public Boolean getstats_desc_hack();
	public Integer getAPIWSPort(Integer id) throws ConfigError;
	public Integer getAPIWSPort() throws ConfigError;
	public Integer getJettyPort(Integer id) throws ConfigError;
	public Integer getJettyPort() throws ConfigError;
	public Integer getListenPort(Integer id) throws ConfigError;
	public Integer getListenPort() throws ConfigError;
	public Boolean getCheckPoint(Integer id) throws ConfigError;
	public Boolean getCheckPoint() throws ConfigError;
	public String getFloodPerm(Integer id) throws ConfigError;
	public String getFloodPerm() throws ConfigError;
	public String getLogIdent(Integer id) throws ConfigError;
	public String getLogIdent() throws ConfigError;
	public String getLogging(Integer id) throws ConfigError;
	public String getLogging() throws ConfigError;
	public String getLogFacility(Integer id) throws ConfigError;
	public String getLogFacility() throws ConfigError;
	public Boolean getTopologyServer(int id) throws ConfigError;
	public Boolean getTopologyServer() throws ConfigError;
	
	
	public void settrack_flows(Integer id, Boolean track_flows);
	public void settrack_flows(Boolean track_flows);
	public void setstats_desc_hack(Integer id, Boolean stats_desc_hack);
	public void setstats_desc_hack(Boolean stats_desc_hack);
	public void setFloodPerm(Integer id, String floodPerm);
	public void setFloodPerm(String floodPerm);
	public void setLogging(Integer id, String logging);
	public void setLogging(String logging);
	public void setLogFacility(Integer id, String logging);
	public void setLogFacility(String logging);
	public void setLogIdent(Integer id, String logging);
	public void setLogIdent(String logging);
	public void setTopologyServer(Integer id, Boolean topo) throws ConfigError;
	public void setTopologyServer(Boolean topo) throws ConfigError;
	public void setListenPort(Integer id, Integer port) throws ConfigError;
	public void setListenPort(Integer port) throws ConfigError;
	public void setAPIWSPort(Integer id, Integer port) throws ConfigError;
	public void setAPIWSPort(Integer port) throws ConfigError;
	public void setJettyPort(Integer id, Integer port) throws ConfigError;
	public void setJettyPort(Integer port) throws ConfigError;
	
	
}
